### PR TITLE
Use Realm queries

### PR DESCRIPTION
**Pull request by KevinBoulongne**
Depends on #243

### CODE REVIEW

File: app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/FolderController.kt [hunk 1]
```diff
@@ -62,12 +63,19 @@ object FolderController {
     private fun MutableRealm?.getFolderQuery(id: String): RealmSingleQuery<Folder> {
         return (this ?: RealmDatabase.mailboxContent).query<Folder>("${Folder::id.name} == '$id'").first()
     }
+
+    fun getCurrentFolder(currentFolderId: String?, defaultRole: FolderRole): Folder? = with(RealmDatabase.mailboxContent) {
+        val folderById = currentFolderId?.let(::getFolder)
+        val folderByRole = query<Folder>("${Folder::_role.name} = '${defaultRole.name}'").first().find()
+        val firstFolder = getFolders().firstOrNull()
+        return folderById ?: folderByRole ?: firstFolder
+    }
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/FolderController.kt:72]: Refactor this query, contains useless cases

File: app/src/main/java/com/infomaniak/mail/data/cache/mailboxInfos/MailboxController.kt [hunk 2]
```diff
@@ -36,6 +36,18 @@ import kotlinx.coroutines.flow.SharedFlow
 object MailboxController {
 
     //region Get data
+    private fun getAllMailboxes(realm: MutableRealm? = null): RealmResults<Mailbox> {
+        return realm.getAllMailboxesQuery().find()
+    }
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/data/cache/mailboxInfos/MailboxController.kt:41]: Why is private ? Is it at least used?

File: app/src/main/java/com/infomaniak/mail/data/cache/mailboxInfos/MailboxController.kt [hunk 3]
```diff
@@ -45,32 +52,34 @@ object MailboxController {
...
     }
 
-    private fun MutableRealm?.getMailboxQuery(objectId: String): RealmSingleQuery<Mailbox> {
-        return (this ?: RealmDatabase.mailboxInfos).query<Mailbox>("${Mailbox::objectId.name} == '$objectId'").first()
+    fun getCurrentMailbox(): Mailbox? = with(RealmDatabase.mailboxInfos) {
+        val checkUserId = "${Mailbox::userId.name} = '${AccountUtils.currentUserId}'"
+        val checkMailboxId = "${Mailbox::mailboxId.name} = '${AccountUtils.currentMailboxId}'"
+        val mailbox = query<Mailbox>("$checkUserId AND $checkMailboxId").first().find()
+        return mailbox ?: query<Mailbox>(checkUserId).first().find()
     }
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/data/cache/mailboxInfos/MailboxController.kt:71]: This method must be generic to be able to reuse it

File: app/src/main/java/com/infomaniak/mail/data/cache/mailboxInfos/MailboxController.kt [hunk 4]
```diff
@@ -45,32 +52,34 @@ object MailboxController {
...
         // Get current data
         Log.d(RealmDatabase.TAG, "Mailboxes: Get current data")
         val realmMailboxes = getMailboxes(AccountUtils.currentUserId)
 
         // Get outdated data
         Log.d(RealmDatabase.TAG, "Mailboxes: Get outdated data")
-        // val deletableMailboxes = MailboxInfoController.getDeletableMailboxes(apiMailboxes)
         val deletableMailboxes = realmMailboxes.filter { realmMailbox ->
             apiMailboxes.none { apiMailbox -> apiMailbox.mailboxId == realmMailbox.mailboxId }
         }
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/data/cache/mailboxInfos/MailboxController.kt:85]: No manual operation if we can use a realm query, and is it really necessary ?

Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/data/cache/mailboxInfos/MailboxController.kt:88]: Can't succeed to do it for now using `IN` realm query 😢

Comment (sirambd) [reply app/src/main/java/com/infomaniak/mail/data/cache/mailboxInfos/MailboxController.kt:88]: To see together then

File: app/src/main/java/com/infomaniak/mail/ui/main/user/SwitchUserAccountsAdapter.kt [hunk 5]
```diff
@@ -101,30 +100,16 @@ class SwitchUserAccountsAdapter(
...
-    }
-
-    private fun List<UiAccount>.sortAccounts(): MutableList<UiAccount> {
-        return filter { it.user.id != AccountUtils.currentUserId }
-            .toMutableList()
-            .apply { this@sortAccounts.find { it.user.id == AccountUtils.currentUserId }?.let { add(0, it) } }
+    private fun collapseAllExceptFirst(newList: List<UiAccount>) {
+        isCollapsed = MutableList(newList.count()) { true }
+        isCollapsed[0] = false
     }
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/ui/main/user/SwitchUserAccountsAdapter.kt:113]: There is no need to do all the processing just to make a collapse, either use the `ViewHolder` or the `Model`.

File: app/src/main/java/com/infomaniak/mail/ui/main/user/SwitchUserViewModel.kt [hunk 6]
```diff
@@ -28,14 +27,30 @@ import com.infomaniak.lib.core.models.user.User
...
-    fun fetchUsersMailboxes(users: List<User>) = viewModelScope.launch(Dispatchers.IO) {
+    fun listenToAccounts(): LiveData<List<UiAccount>> = liveData(Dispatchers.IO) {
+
+        val users = AccountUtils.getAllUsersSync()
+
+        updateMailboxes(users)
+
+        emitSource(MailboxController.getAllMailboxesAsync().map { mailboxes ->
+            users.map { user ->
+                UiAccount(user, mailboxes.list.filter { it.userId == user.id }.sortMailboxes())
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/ui/main/user/SwitchUserViewModel.kt:48]: Use a realm query for this especially with realm you can also do the `sort` easily

File: app/src/main/java/com/infomaniak/mail/data/cache/mailboxInfos/MailboxController.kt (hunk 4 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/data/cache/mailboxInfos/MailboxController.kt]: Done.

File: app/src/main/java/com/infomaniak/mail/ui/main/user/SwitchUserViewModel.kt (hunk 6 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/user/SwitchUserViewModel.kt]: For the `filter`, I'm not sure if we can do it with Realm only.

**Review by sirambd — APPROVED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/user/SwitchUserAccountsAdapter.kt (hunk 5 above)
Comment (sirambd) [reply app/src/main/java/com/infomaniak/mail/ui/main/user/SwitchUserAccountsAdapter.kt:113]: Fix in a new PR

**Pull request merged**